### PR TITLE
[MHLO] Init MHLO basic op conversion

### PR DESCRIPTION
**Pull request by Vremold**
See  RFC #999 
Co-authored-by: Bairen Yi <yibairen.byron@bytedance.com>
Co-authored-by: Jiawei Wu <xremold@gmail.com>
Co-authored-by: Tianyou Guo <tianyou.gty@alibaba-inc.com>
Co-authored-by: Xu Yan <yancey.yx@alibaba-inc.com>
Co-authored-by: Ziheng Jiang <ziheng.jiang@bytedance.com>

### CODE REVIEW

**Comment by Vremold** [comment]
This is a follow-up of PR #1083.
Ready for review. @silvasean @powderluv @fortianyou .Thanks.

File: lib/Conversion/TorchToMhlo/MhloLegalizeUtils.h
Comment (silvasean) [inline lib/Conversion/TorchToMhlo/MhloLegalizeUtils.h:13]: nit: remove `// from @llvm-project` comments

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (silvasean) [inline lib/Conversion/TorchToMhlo/BasicOp.cpp:170]: does this support dynamic shape -- is there a plan to support it? cc @fortianyou

**Review by silvasean — APPROVED** [review]
Overall LGTM. Will let folks iterate on the specific patterns if it hasn't already happened elsewhere.

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (fortianyou) [inline lib/Conversion/TorchToMhlo/BasicOp.cpp:72]: Nit: TOSA -> MHLO

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (fortianyou) [inline lib/Conversion/TorchToMhlo/BasicOp.cpp:159]: Nit: Can we extend the `torchScalarToMhloTensor` to convert also scalar variable but not only constant? Because `add/sub/mul/div` the second operand is not guaranteed to be a constant.

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (fortianyou) [inline lib/Conversion/TorchToMhlo/BasicOp.cpp:179]: ditto, `alpha` can be a variable.

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (fortianyou) [reply lib/Conversion/TorchToMhlo/BasicOp.cpp:170]: Yes, the implicit broadcast might introduce dynamic shapes here. The current implementation might not handle broadcasts correctly if the output shape is unknown. But I think it's ok for now since this is an initial setup; we can fix it in another PR.

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (fortianyou) [inline lib/Conversion/TorchToMhlo/BasicOp.cpp:152]: Will the bool type pass this check?

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (fortianyou) [inline lib/Conversion/TorchToMhlo/BasicOp.cpp:604]: Nit: Update this comment

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (fortianyou) [inline lib/Conversion/TorchToMhlo/BasicOp.cpp:665]: Nit: a better place to convert AtenErfOp is the decomposition pass.

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (fortianyou) [inline lib/Conversion/TorchToMhlo/BasicOp.cpp:851]: ditto

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (fortianyou) [inline lib/Conversion/TorchToMhlo/BasicOp.cpp:740]: ditto

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (Vremold) [reply lib/Conversion/TorchToMhlo/BasicOp.cpp:72]: This part's implementation largely borrows from Torch-to-TOSA lowering. So we marked "TOSA" here.

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (fortianyou) [reply lib/Conversion/TorchToMhlo/BasicOp.cpp:170]: It would be awesome if we could fix this in this PR.

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (Vremold) [reply lib/Conversion/TorchToMhlo/BasicOp.cpp:152]: Yes.

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (Vremold) [reply lib/Conversion/TorchToMhlo/BasicOp.cpp:159]: Yes. Supporting for variable alpha or none-constant rhs in add/sub/mul/div will introduce many code changes. Since it's just a setup, we'd like to support it later.

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (Vremold) [reply lib/Conversion/TorchToMhlo/BasicOp.cpp:179]: ditto

**Review by fortianyou — APPROVED** [review]
LGTM

File: lib/Conversion/TorchToMhlo/BasicOp.cpp
Comment (Vremold) [reply lib/Conversion/TorchToMhlo/BasicOp.cpp]: We have rewritten these codes and provide different APIs to support dynamic shape. We also reimplement many ops‘ conversion patterns in case of dynamic shaped inputs. The change is kind of huge, and probably need a second review. Thanks!

**Comment by Vremold** [comment]
Merged! Thanks @silvasean @fortianyou

**Pull request merged**
